src/mds/MDLog.h

```diff
@@ -91,16 +91,16 @@ class MDLog {
       flush();
   }
 
-  LogSegment *peek_current_segment() {
+  LogSegmentRef peek_current_segment() {
     return segments.empty() ? NULL : segments.rbegin()->second;
   }
 
-  LogSegment *get_current_segment() { 
+  LogSegmentRef get_current_segment() {   //return by value as it being passes by threads..
     ceph_assert(!segments.empty());
     return segments.rbegin()->second;
   }
 
-  LogSegment *get_segment(LogSegment::seq_t seq) {
+  LogSegmentRef get_segment(LogSegment::seq_t seq) { //return by & ?
```

Comment:
I think we actually want to LogSegmentRef& for all three of these functions. (Certainly, they and get_oldest_segment below should match).

Reason being, when we grab a segment reference, it should be protected from deletion while our thread is operating[1]. What we're worried about is when we store the LogSegmentRef into a Context or event and it gets picked up later on, and that store of course increments the internal refcounter for us. While we're in the thread that did the initial lookup, the map's reference is good enough.

[1]: Almost everything in the MDS happens under the big MDS lock, and there are only 3 real threads: 1) main dispatch thread, 2) Finisher thread for handling replies (which grabs the big MDS lock for basically everything), and 3) the logging thread, which avoids the big MDS lock as best it can. So the LogSegment stuff is one of the few cases where we don't work under the big MDS lock all the time, but there's still plenty of coordination going on around moving these LogSegments between different states.

Reply:
thanks for the details :+1: 
done